src/main/groovy/com/morpheusdata/upcloud/sync/VirtualMachinesSync.groovy

```diff
@@ -0,0 +1,377 @@
+                def addConfig = [account:cloud.account, externalId:cloudItem.uuid, name:cloudItem.title, sshUsername:'root',
+                     apiKey:java.util.UUID.randomUUID(), status:'provisioned', hostname:cloudItem.hostname, poweredOn:(cloudItem.state == 'started'),
+                     powerState:(cloudItem.state == 'started' ? 'on' : 'off'), computeServerType:serverType, provision:false,
+                     singleTenant:true, zone:cloud, lvmEnabled:false, managed:false, discovered:true, serverType:'unmanaged'
+                ]
+                def addCapacityConfig = [maxCores:(cloudItem.'core_number' ?: 1), maxMemory:(cloudItem['memory_amount']?.toLong()*ComputeUtility.ONE_MEGABYTE),
+                     maxStorage:0, usedStorage:0
+                ]
+                def serverResults = UpcloudComputeUtility.getServerDetail(authConfig, cloudItem.uuid)
+
+                if(serverResults.success == true && serverResults.server) {
+                    //stats and ip address info
+                    if(serverResults.server.vnc == 'on') {
+                        addConfig.consoleType = 'vnc'
+                        addConfig.consoleHost = cloudItem.'vnc_host'
+                        addConfig.consolePassword = cloudItem.'vnc_password'
+                        addConfig.consolePort = cloudItem.'vnc_port'
+                    }
+                    //ip addresses
+                    if(serverResults.networks) {
+                        def publicIp = serverResults.networks.find{ nic -> nic.family == 'IPv4' && nic.access == 'public' }
+                        def privateIp = serverResults.networks.find{ nic -> nic.family == 'IPv4' && nic.access == 'utility' }
+                        if(privateIp)
+                            addConfig.internalIp = privateIp.address
+                        if(publicIp)
+                            addConfig.externalIp = publicIp.address
+                        addConfig.sshHost = addConfig.externalIp ?: addConfig.internalIp
+                    }
+                    //volumes
+                    if(serverResults.volumes) {
+                        def maxStorage = serverResults.volumes.sum{ volume -> (volume.size ?: 0) }
+                        addCapacityConfig.maxStorage = maxStorage * ComputeUtility.ONE_GIGABYTE
+                    }
+                    //plan
+                    ServicePlan servicePlan = findServicePlanMatch(servicePlans, serverResults.server)
+                    if(servicePlan) {
+                        addConfig.plan = servicePlan
+                    }
+                }
+
+                def add = new ComputeServer(addConfig)
+                addCapacityConfig.server = add
+                add.capacityInfo = new ComputeCapacityInfo()(addCapacityConfig)
+
+                morpheusContext.async.computeServer.create(add)
+                if(serverResults?.volumes)
+                    cacheVirtualMachineVolumes(cloud, add, serverResults.volumes)
+
```

Comment:
Lines 127 to 129 should be done similar to 270 to 273 using bulk create and cache the volumes after bulk create.